libyul/optimiser/DataFlowAnalyzer.cpp

```diff
@@ -280,7 +298,6 @@ void DataFlowAnalyzer::pushScope(bool _functionScope)
 
 void DataFlowAnalyzer::popScope()
 {
-	clearValues(std::move(m_variableScopes.back().variables));
```

Comment:
This here is the single most serious change in the PR.
``clearValues`` is among the most expensive operations, especially since it has to clear back-references.
However, while clearing back-references is absolutely necessary e.g. for assignments, at the end of scopes we don't really have to care that much (joining knowledge on branches is handled differently). The only downside of not clearing here is that the ``DataFlowAnalyzer`` might report matches that are no longer in scope. However, all users/derived classes of ``DataFlowAnalyzer`` already had an additional assertion ``inScope(var)``, whenever a reported match was about to be used. These assertions are just turned into conditions in this PR, allowing us to skip the clearing here.

It's still questionable, of course, but this is the single most effective change here.

Reply:
What if we don't clear back references here, but instead just remove the variables? The upside of clearing at the end of a scope is also that the data structures remain smaller.

Reply:
Yes, I wondered as well, but haven't tried yet.

Reply:
I'm also in general not yet 1000% sure that it's true that not clearing and instead checking on the user side doesn't have any unwanted side-effects...

Reply:
```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m40,954s
user    0m40,611s
sys     0m0,238s
```
If I add the following here:
```
       for (auto const& name: m_variableScopes.back().variables)
       {
               m_value.erase(name);
               m_references.erase(name);
       }
```


vs.

```
$ time { for i in $(seq 10); do build/test/tools/isoltest -t sem*/via* --optimize &> /dev/null ; done;}

real    0m40,649s
user    0m40,308s
sys     0m0,241s
```
without adding it.

So no measurable difference - so I'll add it.

Reply:
(for comparison: current develop is still at around 47 seconds)

Reply:
It looks fine to me but I don't have a deep knowledge about the component. I think we should have a review from @chriseth or @hrkrshnn on this specific question.

Reply:
Not so sure about the effects. Maybe this will have an effect when the ast is not disambiguated. We saw earlier that DataFlowAnalyzer is works well with non-disambiguated code. 

One example of this situation is:

```
  {
      {
          let a := 1
          let b := 2
          mstore(a, b)
      }
      {
          let a := 1
          let b := 3
          let bug := mload(a)
      }
  }
// ----
// step: loadResolver
//
// {
//     {
//         let a := 1
//         let b := 2
//         mstore(a, b)
//     }
//     {
//         let a := 1
//         let b := 3
//         let bug := b
//     }
// }
```

(The output is after removing disambiguator before loadresolver. [Here](https://gist.github.com/hrkrshnn/55f5303e4b6b2f1183818b8eb3eb4560) is the diff)

Reply:
That might very well be - but even if so, I wouldn't say it's a problem - we generally can can assume that sources are disambiguated and we should be able to rely on them being disambiguated. The fact that we had a bug in that once and were lucky that we were not strongly hit by it, is not a good argument for dropping the assumption in general :-).

Reply:
By the way: in general I don't think it works with disambiguated sources - the bug we had just only introduced name conflicts in very specific circumstances, which luckily never resulted in bad code, but in general I'd still expect non-disambiguated code to break anyways.

Reply:
But yeah, it's a good catch, I think you're right, non-disambiguated sources would indeed have problems with this change... not sure that should stop us, though.